### PR TITLE
RFC: Check lvalue of Array_get in ast_check

### PR DESCRIPTION
**Pull request by fangyi-zhou**
Summary: See #8314 for context
In the Array_get lvalue, we only checked whether the array being modified
is a class const member and raised a parse error, so we were able to
reject code like this

```
class Foo {
    const varray<int> DATA = varray[42];
    public function foo(): void {
        Foo::DATA[] = -1;
    }
}
```

In order for an Array_get lvalue to be valid, the array must be a valid
lvalue. This commit addresses this issue by adding this requirement.

For reference, the following code raises runtime fatal in PHP
(https://3v4l.org/bWGMO)

```
<?php

const FOO = array(42);
FOO[] = -1;
```

but was accepted by HackC (tested locally)

Side Note: We allow return values from functions to be a valid lvalue
for the moment. There probably needs to be checks by the HHVM runtime
elsewhere in addition to this parser change. This is necessary because
Hack Collections don't have value semantics and we can construct code
like to exploit this.

Luckily PHP doesn't accept lvalue that is a conditional expression -
this is possible in C:

Valid C Code:
```
int foo(bool cond) {
    int a = 0 , b = 0;
    (cond ? a : b) = 42;
    return a;
}
```

Whereas its PHP equivalent is parse error: https://3v4l.org/Yk0Ac
```
function foo($cond) {
    ($cond ? $a : $b) = 42;
    return $a;
}
```

### CODE REVIEW

**Comment by fangyi-zhou** [comment]
Could I get someone from typechecker team to look at it?

**Comment by fangyi-zhou** [comment]
#m pinguping

File: hphp/system/php/spl/datastructures/SplObjectStorage.php [hunk 1]
```diff
@@ -364,7 +364,8 @@ public function unserialize($serialized) {
    * @return     mixed   No value is returned.
    */
   public function setInfo($data) {
-    current(&$this->__storage)['inf'] = $data;;
+    $storage = current(&$this->__storage);
```
Comment (fredemmott) [inline hphp/system/php/spl/datastructures/SplObjectStorage.php]: Was this already broken, or is there a way to make this syntax work in PHP Code?

Comment (fangyi-zhou) [reply hphp/system/php/spl/datastructures/SplObjectStorage.php]: I reenabled the syntax in the second commit but forgot to revert this change. Will revert this change.

**Comment by fangyi-zhou** [comment]
Rebased and reverted the change in systemlib

**Comment by fangyi-zhou** [comment]
#m pingping

**Comment by fredemmott** [comment]
This is a smell, but:

```Hack
class :foo {
  attribute Map<string, string> bar;

  public function doStuff(): void {
    $this->:bar['baz'] = 'herpderp'; // incorrectly raises lvalue error
  }
}
```

**Pull request closed without merging**